sdk/go/common/util/gitutil/git.go

```diff
@@ -237,7 +245,45 @@ func getMatchedGroupsFromRegex(regex *regexp.Regexp, remoteURL string) map[strin
 	return groups
 }
 
-func parseAuthURL(remoteURL string) (string, *http.BasicAuth, error) {
+// sshKeys memoizes keys we've loaded for given host URLs, to avoid needing to
+// re-fetch public keys.
+var sshKeys = map[string]transport.AuthMethod{}
+
+// urlIsSSH returns true if the given URL is of the form `git@...`, `ssh://git@...`,
+// etc.
+func urlIsSSH(remoteURL string) (*transport.Endpoint, bool) {
+	endpoint, err := transport.NewEndpoint(remoteURL)
+	if err != nil {
+		return nil, false
+	}
+	return endpoint, endpoint.Protocol == "ssh"
+}
+
+// parseAuthURL extracts HTTP basic auth parameters if provided in the URL.
+//
+// If the URL uses SSH, the user's SSH configuration is parsed and relevant
+// public keys are returned for authentication.
+//
+// sshConfig is provided for testing only. When nil it will read SSH
+// configuration from disk.
+func parseAuthURL(remoteURL string, sshConfig *sshConfig) (string, transport.AuthMethod, error) {
+	if endpoint, ok := urlIsSSH(remoteURL); ok {
+		// See if we've encountered this host before; if yes, use the existing key.
+		if existing, ok := sshKeys[endpoint.Host]; ok {
+			return remoteURL, existing, nil
+		}
+
+		auth, err := getSSHPublicKeys(endpoint.User, endpoint.Host, sshConfig)
+		if err != nil {
+			return "", nil, fmt.Errorf("no valid SSH key file found for Git URL %s: %w", remoteURL, err)
+		}
+
+		// Memoize and return the key.
+		sshKeys[endpoint.Host] = auth
+		return remoteURL, auth, nil
+	}
+
+	// For non-SSH URLs, see if there is basic auth info.
 	u, err := url.Parse(remoteURL)
```

Comment:
This is redundant. `transport.NewEndpoint` will attempt `url.Parse` and put the result in [Endpoint.User](https://pkg.go.dev/github.com/go-git/go-git/v5/plumbing/transport#Endpoint.User](https://pkg.go.dev/github.com/go-git/go-git/v5/plumbing/transport#Endpoint.User) and Endpoint.Password.

Reply:
Yep, cleaned up a similar redundant `url.Parse` in the `https` case above. I didn't want to touch too much code while I was in here but fortunately there's a good test suite so I'm comfortable with this cleanup.